File: src/Infrastructure/NetFusion.Rest.Client/Settings/QueryString.cs
using System;
using System.Collections.Generic;
using System.Collections.ObjectModel;

namespace NetFusion.Rest.Client.Settings
{
    /// <summary>
    /// Used to specify query string parameters used when making a request.
    /// </summary>
    public class QueryString
    {
        /// <summary>
        /// The collection of parameters.
        /// </summary>
        public IReadOnlyDictionary<string, string> Params { get; } 

        private readonly IDictionary<string, string> _params;

        public QueryString()
        {
			_params = new Dictionary<string, string>();
			Params = new ReadOnlyDictionary<string, string>(_params);
        }

        /// <summary>
        /// Adds a query string parameter name and value to be added to query string.
        /// </summary>
        /// <returns>Reference to query string instance.</returns>
        /// <param name="name">Name of the query string parameter.</param>
        /// <param name="value">Value of the query string parameter.</param>
        public QueryString AddParam(string name, string value)
        {
			if (string.IsNullOrWhiteSpace(name))
				throw new ArgumentException("Query string parameter name not specified.", nameof(name));

			if (string.IsNullOrWhiteSpace(value))
				throw new ArgumentException("Query string parameter value not specified.", nameof(value));

            _params[name] = value;
            return this;
        }

        internal QueryString GetMerged(QueryString queryString)
        {
            var mergedQuery = new QueryString();

			// Add all the params from this query object to the new merged version.
			foreach (var param in Params)
			{
				mergedQuery._params[param.Key] = param.Value;
			}

			// Override or add any param values being merged.
			foreach (var param in queryString.Params)
			{
				mergedQuery._params[param.Key] = param.Value;
			}

            return mergedQuery;
        }
    }
}

File: netfusion/tutorial/Client/src/app/common/rest-client/QueryString.ts
import { forOwn, clone } from "lodash";
import { MatPaginator, MatSort } from '@angular/material';

/**
 * Settings class used to specify query string parameters.
 */
export class QueryString {
    private values: {[name:string]: string} = {};

    /**
     * Adds a query string parameter to the settings.
     * 
     * @param name The name of the query string parameter.
     * @param value The value of the query string parameter.
     */
    public addParam(name: string, value: string): QueryString {
        this.values[name] = value;
        return this;
    }

    public addParams(values: {[name:string]:string}): QueryString {
        forOwn(values, (v, n) => this.values[n] = v)
        return this;
    }

    /**
     * Returns a copy of the query-string name/value pairs.
     */
    public getParams(): {[name:string]:string} {
        return clone(this.values);
    }

    /**
     * Creates a new QueryString instance consisting of the specified settings 
     * merged into the settings of the current instance.
     * 
     * @param queryString The query string settings to be merged.
     */
    public getMerged(queryString: QueryString): QueryString {
        let mergedQuery = new QueryString

        mergedQuery.addParams(this.values);
        mergedQuery.addParams(queryString.values);
        return mergedQuery;
    }
}

export class QueryInfo {
    pageIndex: number;
    pageSize: number;

    sortBy?: string;
    sortDirection?: string;
    filerBy?: {[field: string]: any}   

    public static pageAndSort(paginator: MatPaginator, sort: MatSort): QueryInfo {
        let queryInfo = new QueryInfo();

        queryInfo.pageIndex = paginator.pageIndex || 0;
        queryInfo.pageSize = paginator.pageSize || 10;
        queryInfo.sortBy = sort.active,
        queryInfo.sortDirection = sort.direction;
    
        return queryInfo;
    }

    public static page(paginator: MatPaginator): QueryInfo {
        let queryInfo = new QueryInfo();

        queryInfo.pageIndex = paginator.pageIndex || 0;
        queryInfo.pageSize = paginator.pageSize || 10; 
    
        return queryInfo;
    }
}